feat: add docs for Post Message action / postWindowMessage (#646)
appsmithorg/appsmith-docs: website/docs/reference/appsmith-framework/widget-actions/post-message.md

- Embedding Appsmith in an iframe of a parent application.

## Signature

```javascript
postWindowMessage(message, targetIframe, targetOrigin)
```

### Arguments

| Argument Name | Description |
| ------------- | ----------- |
| **message** | This is the message to send to the target iframe or window. Most JavaScript values are acceptable here, except `null` and `undefined`. (Default: `""`) |
| **targetIframe** | This is the window to which you want to send the message. If its value is `"window"`, you are sending a message to the parent application’s window (where Appsmith is embedded). If you want to send a message to an iframe within Appsmith, enter the name of the iframe here. (Default: `"window"`) |
| **targetOrigin** | This is the URL to which you can send messages. The default value of "`*`" means the message can be sent to any URL. If you want to limit sending messages to only the parent application (in which Appsmith is embedded), enter the URL of the parent application here. (Default: `"*"`) |

To see examples of the **Post message** action, take a look at the [sample app](https://app.appsmith.com/applications/61f3d1949d6d6a6720c98681/pages/61f3d1949d6d6a6720c98684).

> 

Reviewer: ok here I recall earlier we called the sub-heading 'Examples', how about we call it 'Types of communication' or 'Levels of communication' and create an image like this one by Rimil (in the doc image, we wont include the WIP content, only the ones that are available today). I think it will help users understand very clearly before diving into each type

<img width="892" alt="Screenshot 2022-12-16 at 17 52 51" src="https://user-images.githubusercontent.com/32227135/208098054-6c87ddbe-ec73-403b-a422-7f3ebbe2633f.png">

We will have to update this image later when the other two communication levels also become available.

Author: great idea, these diagrams were extremely helpful

Reviewer: image shared by Rimil
![Untitled-2022-11-16-0828](https://user-images.githubusercontent.com/32227135/209400859-0d73904e-e4a6-4387-9178-62ae59bfd3c4.png)

## Iframe widget in Appsmith app

> In this scenario, you are building an Appsmith that has an Iframe widget called `Iframe1`. There is an external page embedded within that Iframe widget, and you are setting up communication between the embedded page and the Iframe in Appsmith.

Reviewer: ```suggestion
In this scenario, you are building an app that has an Iframe widget called `Iframe1`. There is an external page embedded within that Iframe widget, and you are setting up communication between the embedded page and the Iframe in Appsmith.
```

# Post Message

> The **Post message** action is used to enable safe cross-origin communication between [Window](https://developer.mozilla.org/en-US/docs/Web/API/Window) objects; for example, between a page and a pop-up that it has created, or between a page and an iframe embedded within it. The Appsmith function `postWindowMessage` works by using the JavaScript function [`postMessage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) so that you can communicate between applications in situations such as:

Author: I made some changes to the suggestion above, I this sounds clearer and shorter than my original :)

> The **Post message** action uses the `postWindowMessage()` function to enable safe cross-origin communication between different [Window](https://developer.mozilla.org/en-US/docs/Web/API/Window) objects; for example, between an app and an iframe embedded within it. This may be useful in situations like:

> 

Author: (Included an alternative in the solution for the comment above)

> - Appsmith being embedded in an iframe of a parent application.

Reviewer: ```suggestion
- Appsmith app is embedded in an iframe of a parent application.
```

also let's move this  down as the second point

- Appsmith using [Iframe](/reference/widgets/iframe/) widgets to embed other applications or pages.

## Signature

```javascript
postWindowMessage(message, targetIframe, targetOrigin)
```

### Arguments

| Argument Name | Description |
| ------------- | ----------- |
| **message** | This is the message to send to the parent application. Most JavaScript values are acceptable here, except `null` and `undefined`. (Default: `""`) |
| **targetIframe** | This is the window to which you want to send the message. If its value is `"window"`, you are sending a message to the parent application’s window (where Appsmith is embedded). If you want to send a message to an iframe within Appsmith, enter the name of the iframe here. (Default: `"window"`) |
| **targetOrigin** | This is the URL to which you can send messages. The default value of "`*`" means the message can be sent to any URL. If you want to limit sending messages to only the parent application (in which Appsmith is embedded), enter the URL of the parent application here. (Default: `"*"`) |

> 

Author: Added! Just as a note though, the sample app is pretty absent of information, I'm not sure how useful it will be in its current state.

## Examples

### Page embedded in Appsmith

> For these examples, imagine you are building an app in Appsmith that has an Iframe widget called `Iframe1` containing an embedded page.

Reviewer: delete this line.  Here we can explain the scenario a bit deeper  - a page or app embedded in an Appsmith iframe widget.......

##### From Appsmith to iframe

> To send a piece of data (perhaps an email address) to that embedded page:

Reviewer: ```suggestion
To send data to an embedded page in an Appsmith IFrame widget:
```

1. Drag and drop an [Input](/reference/widgets/input/) widget and a [Button](/reference/widgets/button) widget onto the canvas.
2. In the button's **onClick** event property, select **Post message**. Set the **Message** to `{{ Input1.text }}` and **Target iframe** to `Iframe1`.
3. On the canvas, enter your piece of data into the input widget, and click the button widget. The iframe has now received the message you sent.

> :::note

Reviewer: Isn't this a necessary step so that users can see the feature in action?

Reviewer: Then based on what you said above, I strongly recommend we avoid putting this in a note. especially because it stands out as important information that you want users to look at. As you are already starting the para with the scenario  "If you want to try this but you need a page to receive your message, ......" I think it is better to leave it out.  that info is too big to be inside a callout note :)

If you want to try this but you need a page to receive your message, set up your iframe with this HTML string in its **srcDoc** property:

```html
<div id="target"></div>

<script>
    window.addEventListener('message', (event) => {
    const tgt = document.querySelector("#target")
        tgt.textContent = event.data
    });
</script>
```

When the iframe receives your message, it puts the message text into the `#target` div.
:::

> ##### From iframe to Appsmith

Author: Updated heading from "**From embedded page to Iframe widget**" for accuracy. Let me know if you have a better idea, I can't think of anything else -- I think this is a bit harder to understand as a reader, but is technically correct.

> ##### From iframe to Appsmith

Reviewer: This sounds ok. But inside the explanation let's make it clearer by saying the below

> To capture and handle a message that was sent to an Appsmith Iframe widget from an entity inside the page embedded in the iframe:

> ##### From iframe to Appsmith

Reviewer: @spark-c could you update the above line before merging? Thanks

### Appsmith app embedded in parent app

In this scenario, you are building an app outside of Appsmith, and you have used an [HTML iframe](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe) element to embed a deployed Appsmith app in that page. Now you are setting up communication between the non-Appsmith parent app and the Appsmith app embedded within it.

#### From Appsmith to parent app

To configure your Appsmith application to send messages to its parent page:

1. Drag and drop a [Button](/reference/widgets/button/) widget and an [Input](/reference/widgets/input/) widget onto the canvas.
2. In the button's properties, configure its **onClick** event to execute the **Post message** action. Set its **Message** field to `{{Input1.text}}`, and make sure that the **Target iframe** field is set to `window`.
3. On the canvas, enter some data in the input widget's field and click the button widget. Appsmith has just emitted a message to its parent window.

> :::tip

Reviewer: ```suggestion
:::note
```
